### PR TITLE
Metrics and Alerting for Metris

### PR DESCRIPTION
**Pull request by grischperl**
**Description**

Changes proposed in this pull request:

- Report general statistics (failed requests, total requests, sent request duration)
- Report error count of getting list with runtimes from KEB
- Report error count of adding subaccount to cache
- Report error count of getting metrics from the cluster
- Report error count of getting kubeconfig from the cluster
- Report error count of sending metrics to EDP
- Report number of clusters scraped
- Report ratio between number of current clusters scraped to previous
- Configure alerts based on those metrics

**Related issue(s)**

#491

### CODE REVIEW

File: components/kyma-metrics-collector/pkg/keb/client.go [hunk 1]
```diff
@@ -42,6 +43,8 @@ func NewClient(config *Config, logger *logrus.Logger) *Client {
 func (c Client) NewRequest() (*http.Request, error) {
 	kebURL, err := url.ParseRequestURI(c.Config.URL)
 	if err != nil {
+		totalRequest.WithLabelValues(fmt.Sprintf("%d", http.StatusBadRequest)).Inc()
```
Comment (k15r) [inline components/kyma-metrics-collector/pkg/keb/client.go:46]: in this case we do not have a request at all, so also no bad request

File: components/kyma-metrics-collector/pkg/keb/client.go [hunk 2]
```diff
@@ -108,13 +114,17 @@ func (c Client) GetRuntimesPerPage(req *http.Request, pageNum int) (*kebruntime.
 	}
 
 	if resp.StatusCode != http.StatusOK {
+		totalRequest.WithLabelValues(fmt.Sprintf("%d", resp.StatusCode)).Inc()
```
Comment (k15r) [inline components/kyma-metrics-collector/pkg/keb/client.go:117]: we only need to do this once, so if you move this to line 115, all other similar calls in this function can be removed

File: components/kyma-metrics-collector/pkg/process/process.go [hunk 3]
```diff
@@ -284,21 +299,26 @@ func (p Process) getRecordWithOldOrNewMetric(identifier int, subAccountID string
...
 	}
+	gardenerErrorCount.Reset()
+	skrErrorCount.Reset()
 	return &record, false, nil
 }
 
 func (p Process) sendEventStreamToEDP(tenant string, payload []byte) error {
 	edpRequest, err := p.EDPClient.NewRequest(tenant)
 	if err != nil {
+		edpErrorCount.WithLabelValues("failed_request", tenant).Inc()
```
Comment (k15r) [inline components/kyma-metrics-collector/pkg/process/process.go:310]: The name of this metric could be improved. 'failed_request' could mean, that the request fails, but in this case it means that creating the request object failed.

File: components/kyma-metrics-collector/pkg/keb/client.go (hunk 2 above)
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/keb/client.go:117]: done

File: components/kyma-metrics-collector/pkg/process/process.go (hunk 3 above)
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/process.go:310]: changed label to "failed_createRequest"

File: components/kyma-metrics-collector/pkg/edp/client.go [hunk 4]
```diff
@@ -52,6 +54,8 @@ func (eClient Client) NewRequest(dataTenant string) (*http.Request, error) {
 
 	req, err := http.NewRequest(http.MethodPost, edpURL, bytes.NewBuffer([]byte{}))
 	if err != nil {
+		totalRequest.WithLabelValues(fmt.Sprintf("%d", http.StatusBadRequest)).Inc()
```
Comment (grischperl) [inline components/kyma-metrics-collector/pkg/edp/client.go:57]: @k15r this shouldn't be a label bad request either? but rather "failed_createRequest" for the same reason as you mentioned for the keb client [here](https://github.com/kyma-project/control-plane/pull/667#discussion_r641636208)

File: components/kyma-metrics-collector/pkg/keb/client.go (hunk 1 above)
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/keb/client.go:46]: changed label to "failed_createRequest"

**Review by k15r — APPROVED** [review]

File: components/kyma-metrics-collector/pkg/process/public_cloud_info.go [hunk 5]
```diff
@@ -44,30 +44,36 @@ func (p Providers) GetFeature(cloudProvider, vmType string) (f *Feature) {
 // LoadPublicCloudSpecs loads string data to Providers object from an env var
 func LoadPublicCloudSpecs(cfg *env.Config) (*Providers, error) {
 	if cfg.PublicCloudSpecs == "" {
+		providerErrorCount.WithLabelValues("spec_missing").Inc()
```
Comment (sayanh) [inline components/kyma-metrics-collector/pkg/process/public_cloud_info.go:47]: `LoadPublicCloudSpecs` happens during the bootup only. If this fails, there is `log.Fatalf`. Hence this is no need to have metrics for all these errors, IMO.

**Review by sayanh — CHANGES_REQUESTED** [review]
In general, we need to have tests for the metrics.

File: resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+{{ include "kyma-metrics-collector.prometheusrule.labels" . | indent 4 }}
+  name: {{ include "kyma-metrics-collector.fullname" . }}
+  namespace: {{ .Values.prometheus.namespace }}
+spec:
+  groups:
+    - name: kmc.rules
```
Comment (sayanh) [inline resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml:12]: (nit): let's use `EDP`, `KMC`, `KEB`, `SKR` consistently all over the places as these acronyms

File: resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+{{ include "kyma-metrics-collector.labels" . | indent 4 }}
+  name: {{ include "kyma-metrics-collector.fullname" . }}
+  namespace: kcp-system
```
Comment (sayanh) [inline resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml:7]: Let's not explicitly use namespace. Helm will install this resource in the namespace where the chart is installed.

File: resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+spec:
+  endpoints:
+   - port: http
+  namespaceSelector:
+    matchNames:
+    - kcp-system
```
Comment (sayanh) [inline resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml:13]: Should be templated

File: resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      rules:
+      - alert: KmcDown
```
Comment (sayanh) [inline resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml:14]:
```suggestion
      - alert: KMCDown
```
(nit): let's use `EDP`, `KMC`, `KEB`, `SKR` consistently all over the places as these acronyms

File: components/kyma-metrics-collector/pkg/edp/client.go [hunk 10]
```diff
@@ -52,6 +54,8 @@ func (eClient Client) NewRequest(dataTenant string) (*http.Request, error) {
 
 	req, err := http.NewRequest(http.MethodPost, edpURL, bytes.NewBuffer([]byte{}))
 	if err != nil {
+		totalRequest.WithLabelValues("failed_createRequest").Inc()
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/edp/client.go:57]: done

File: components/kyma-metrics-collector/pkg/edp/metrics.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	Subsystem = "edp"
+)
+
+var (
+	totalRequest = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: Namespace,
+			Subsystem: Subsystem,
+			Name:      "request_total",
+			Help:      "Total number of requests.",
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/edp/metrics.go:19]: done

File: components/kyma-metrics-collector/pkg/keb/metrics.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+		},
+		[]string{"status"},
+	)
+
+	sentRequestDuration = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Namespace: Namespace,
+			Subsystem: Subsystem,
+			Name:      "request_duration_seconds",
+			Help:      "Duration of HTTP request to EDP in seconds.",
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/keb/metrics.go:29]: done

File: components/kyma-metrics-collector/pkg/keb/metrics.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	Subsystem = "keb"
+)
+
+var (
+	totalRequest = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: Namespace,
+			Subsystem: Subsystem,
+			Name:      "request_total",
+			Help:      "Total number of requests.",
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/keb/metrics.go:19]: done

File: components/kyma-metrics-collector/pkg/process/metrics.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+		},
+		[]string{"requestURI"},
+	)
+
+	providerErrorCount = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: Namespace,
+			Subsystem: "public_cloud",
+			Name:      ErrorCountName,
+			Help:      "Number of continuous errors of sth sth.",
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/metrics.go:79]: I meant to correct this later, fixed it now

File: components/kyma-metrics-collector/pkg/process/process.go [hunk 15]
```diff
@@ -87,46 +89,54 @@ func (p Process) generateRecordWithMetrics(identifier int, subAccountID string)
 	var shoot *gardenerv1beta1.Shoot
 	shoot, err = p.ShootClient.Get(ctx, shootName)
 	if err != nil {
+		gardenerErrorCount.WithLabelValues("shoot", shootName, subAccountID).Inc()
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/process.go:92]: done

File: components/kyma-metrics-collector/pkg/process/process.go [hunk 16]
```diff
@@ -87,46 +89,54 @@ func (p Process) generateRecordWithMetrics(identifier int, subAccountID string)
...
 		return
 	}
 
 	// Get nodes dynamic client
 	nodesClient, err := p.NodeConfig.NewClient(record.KubeConfig)
 	if err != nil {
+		skrErrorCount.WithLabelValues("nodes_client", subAccountID).Inc()
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/process.go:99]: done

File: components/kyma-metrics-collector/pkg/process/process.go [hunk 17]
```diff
@@ -87,46 +89,54 @@ func (p Process) generateRecordWithMetrics(identifier int, subAccountID string)
...
 		return
 	}
 
 	// Get nodes
 	var nodes *corev1.NodeList
 	nodes, err = nodesClient.List(ctx)
 	if err != nil {
+		skrErrorCount.WithLabelValues("nodes", subAccountID).Inc()
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/process.go:107]: done

File: components/kyma-metrics-collector/pkg/process/public_cloud_info.go (hunk 5 above)
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/process/public_cloud_info.go:47]: I see! Removed

File: components/kyma-metrics-collector/pkg/edp/client.go [hunk 18]
```diff
@@ -94,6 +105,8 @@ func (eClient Client) Send(req *http.Request, payload []byte) (*http.Response, e
 	})
 
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/edp/client.go:107]: done

File: components/kyma-metrics-collector/pkg/edp/client.go [hunk 19]
```diff
@@ -80,12 +87,16 @@ func (eClient Client) Send(req *http.Request, payload []byte) (*http.Response, e
 		req.Body = ioutil.NopCloser(bytes.NewReader(payload))
 		resp, err = eClient.HttpClient.Do(req)
 		if err != nil {
+			totalRequest.WithLabelValues(fmt.Sprintf("%d", resp.StatusCode)).Inc()
```
[the remark this answers is not in the document]
Comment (grischperl) [reply components/kyma-metrics-collector/pkg/edp/client.go:90]: done

File: resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml (hunk 9 above)
Comment (grischperl) [reply resources/kcp/charts/kyma-metrics-collector/templates/prometheusrule.yaml:14]: done

File: resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml (hunk 7 above)
Comment (grischperl) [reply resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml:7]: done

File: resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml (hunk 8 above)
Comment (grischperl) [reply resources/kcp/charts/kyma-metrics-collector/templates/service-monitor.yaml:13]: done

**Review by sayanh — APPROVED** [review]

**Review by alexandra-simeonova — APPROVED** [review]
👍

**Pull request merged**